File: string_search/rabin_karp.cpp
#include<bits/stdc++.h>
using namespace std;

#define MAX_CHAR 256
 
void search(char pattern[], char text[], int modulo_prime = 101) {

    int M = strlen(pattern);
    int N = strlen(text);
    int i, j;
    int pattern_hash = 0; 
    int text_hash = 0; 
    int power_hash = 1;
 
    for (i = 0; i < M-1; i++) {
        power_hash = (power_hash*MAX_CHAR)%modulo_prime;
    }

    for (i = 0; i < M; i++) {
        pattern_hash = (MAX_CHAR*pattern_hash + pattern[i])%modulo_prime;
        text_hash = (MAX_CHAR*text_hash + text[i])%modulo_prime;
    }
 
    for (i = 0; i <= N - M; i++) {
        // Check the hash values of current window of text
        // and pattern. 
        if (pattern_hash == text_hash) {
            for (j = 0; j < M; j++) {
                if (text[i+j] != pattern[j]) {
                    break;
                }
            }
            if (j == M) {
                cout << "Pattern found at index: " << i << "\n";
            }
        }
 
        // Calculate hash value for next window of text: Remove
        // leading digit, add trailing digit
        if (i < N-M) {
            text_hash = (MAX_CHAR*(text_hash - text[i]*power_hash) + text[i+M])%modulo_prime;
            if (text_hash < 0) {
                text_hash = (text_hash + modulo_prime);
            }
        }
    }
}

File: strings/string_search/rabin_karp/python/rabin-karp.py
# Function requires the following arguments:
# text,
# pattern,
# d = number of characters in input alphabet,
# q = a prime number


def rabin_karp_algorithm(text, pattern, d, q):
    n = len(text)
    m = len(pattern)
    h = pow(d, m-1) % q
    p = 0                   # hash value of the pattern
    t = 0                   # hash value of the text
    result = []
    # calculate the hash values for pattern and text
    for i in range(m):
        p = (d * p + ord(pattern[i])) % q
        t = (d * t + ord(text[i])) % q

    # Slide pattern over text
    for s in range(n - m + 1):
        # Check the hash values of the current window of text and pattern
        if p == t:
            # Check for characters one by one
            match = True
            for i in range(m):
                if pattern[i] != text[s+i]:
                    match = False
                    break
            if match:
                result = result + [s]
        # Calculate hash value for next window of text: Remove
        # leading digit, add trailing digit
        if s < n - m:
            t = (t - h * ord(text[s])) % q  # remove letter s
            t = (t * d + ord(text[s+m])) % q  # add letter s+m
            t = (t + q) % q  # make sure that t >= 0
    return "Found pattern at " + str(result) + "."


print(rabin_karp_algorithm("Hello World", "ll", 256, 11))
print(rabin_karp_algorithm("12121212", "12", 256, 11))

